### DISCUSSION TITLE
PHP 8+ compatibility

### DISCUSSION THREAD

**Discussion by Gitinou** [postfixadmin/postfixadmin#716] [Q&A]
Hi all, 
Correct me if I'm wrong, but in my experience PFA has become incompatible with recent PHP releases (8 but also 7.4+ which I tried), due to lack of some mbstring functions, even with --enabled-mbstring ? e.g mb_split doesn't exist any more, whereas mb_convert_encoding still does. An I right, and if yes, what would be a workaround ??

**Comment by DavidGoodwin** [comment]
Hi - What are you referring to when you say '--enabled-mbstring' ?  Which version of Postfixadmin are you referring to?

The github build seems happy on PHP8 ...

https://www.php.net/mb_split -> does exist in PHP 8

**Comment by Gitinou** [comment]
Version is 3.3.13, latest, I believe.

--enable-mbstring is a build switch when compiling PHP. It ensures the bundled mbstring code is included in the Makefile. I mentioned it here just in case someone would suggest it, as it is the main (usual) cause for mbstring not working.

I know that the function is supposed to be included (said php.net), but it appears some of them have been suppressed (https://nahan.fr/migration-de-code-vers-php-8-mbstring-partie-2 , sorry, in french). Some are still there :

php testmb.php 
mb_convert_encoding : **Y**
mb_split : **N**

Tried on several PHP8.2, and a 7.4.9 which I happened to have here...

**Reply by Gitinou** [reply]
Just so you know : tested OK under PHP 7.3 : 

php testmb.php 
mb_convert_encoding : **Y**
mb_split : **Y**

php --version
PHP 7.3.28 (cli) (built: Jun  7 2021 01:37:27) ( ZTS )

**Comment by UlyssesRamosju** [comment]
hope you can figure out a solution soon!thanks

**Reply by Gitinou** [reply]
Hi,
No  feedback ? Having such a heavy dependence to this library seems... a lot, compared to what PFA does. And PHP7 is deprecated, so PFA will be as well very soon. Sad.

**Reply by DavidGoodwin** [reply]
@Gitinou I run postfixadmin under PHP 8.1 and don't see a problem. 

If you're missing mb_ functions, it's probably due to you not having necessary packages installed (e.g. php8.1-mbstring). 

If you're compiling PHP yourself from source, I'm afraid I can't help you - as it's been ~10 years since I did that regularly.

See e.g.

https://www.php.net/mb_split - mb_split exists for PHP8 
https://www.php.net/mb_convert_encoding fine for PHP8 ...

**Comment by Gitinou** [comment]
Hi all, 

I found it : it is not documented in the said functions, but some of the mb_ functions may be disabled by a configuration switch, when compiling PHP : --disable-mbregex ; and you HAVE to define this if you don't want to link to a mandatory external library "oniguruma", which is expected to be required when dealing with regexp + japanese or chinese symbols. BUT if you choose to not-enable this, you also lose some functions... like mb_split() ! Which you cannot figure is an eastern-regexep handler. (?)

Long story short, the function is back if you compile PHP with oniguruma library. Everything back to normal.

Thanks for reading !
